### ISSUE TITLE
Issue with Samba share and Windows ACLs

### ISSUE THREAD

**Issue by puyolsmash** [trapexit/mergerfs#765]
General description
===================
I'm running an OMV5 server with the following characteristics:

•	Using SnapRAID to back up my disks
•	Using mergerfs to create 1 big volume
•	Configured samba as an AD domain member (https://wiki.samba.org/index.php/Setting_up_Samba_as_a_Domain_Member)
•	Configured the samba share using Windows ACLs (https://wiki.samba.org/index.php/Setting_up_a_Share_Using_Windows_ACLs) (permissions are set in Windows, not in Linux)

The issue I’m facing is when I use mergerfs to merge my disks to one big volume: when my samba share points to the mergerfs mount point, I am unable to access my share with any other user than the administrator account. It seems that mergerfs does not “pass through” the Windows ACLs but only the linux permissions. I assume the reason I access the share with my admin account is because it’s mapped in Samba to my linux root user. A Samba share that does not point to a mergerfs volume does not have this issue; I can access the share based on the file permissions I set in Windows. Strangely enough it works fine with mhddfs, but this slows down write speed significantly…

Expected behavior
=================
Using a Samba share that points to a mergerfs mount point I expect the permissions I set on the share in Windows be taken into consideration.

Actual behavior
===============
The share is not accessible by any of the AD users except for the administrator account which might be caused by the user mapping in Samba.

**Precise** steps to reproduce the behavior
===========================================
1: Create a mergerfs volume. I used the following in my fstab file:
/srv/dev-disk-by-label-data* /mnt/data/ fuse.mergerfs defaults,allow_other,use_ino 0 0

2: Configure Samba as a domain member:
https://wiki.samba.org/index.php/Setting_up_Samba_as_a_Domain_Member

3: Configured the samba share using Windows ACLs:
https://wiki.samba.org/index.php/Setting_up_a_Share_Using_Windows_ACLs
In Windows, I set my share permissions to Everyone (full control) and the folder permissions to:
- Domain Admins (full control)
- Domain Users (modify)

4: On Windows, connect to the Samba share and authenticate with a Domain Users member. Either you get "You do not have permission to access \\SHARE contact your network administrator to request access" or it just does nothing when trying to enter the folder.

System information
==================
root@omv:~# mergerfs -V
mergerfs version: 2.29.0
FUSE library version: 2.9.7-mergerfs_2.29.0
fusermount version: 2.9.7-mergerfs_2.29.0
using FUSE kernel interface version 7.31

root@omv:~# cat /etc/fstab
UUID=1a9c74cf-77a6-42fc-a064-f8aa38e137c1 /               ext4    errors=remount-ro 0       1
UUID=b42a033c-2b99-440a-bec6-5a1c020eb3a4 none            swap    sw              0       0
/dev/sr0        /media/cdrom0   udf,iso9660 user,noauto     0       0

#mergerfs pool
#/srv/dev-disk-by-label-data* /mnt/data/ fuse.mergerfs defaults,allow_other,use_ino 0 0
#/home/test0* /home/merge/ fuse.mergerfs direct_io,defaults,allow_other,fsname=mergerfs 0 0

#mhddfs pool (seems to be the only union mount working with windows acl, but slow, old and buggy...)
mhddfs#/srv/dev-disk-by-label-data1,/srv/dev-disk-by-label-data2 /mnt/data fuse defaults,allow_other,nonempty 0 0

/dev/disk/by-label/parity1              /srv/dev-disk-by-label-parity1  ext4    defaults,nofail,user_xattr,noexec,usrjquota=aquota.user,grpjquota=aquota.group,jqfmt=vfsv0,acl  0 2
/dev/disk/by-label/data1                /srv/dev-disk-by-label-data1    ext4    defaults,nofail,user_xattr,noexec,usrjquota=aquota.user,grpjquota=aquota.group,jqfmt=vfsv0,acl  0 2
/dev/disk/by-label/data2                /srv/dev-disk-by-label-data2    ext4    defaults,nofail,user_xattr,noexec,usrjquota=aquota.user,grpjquota=aquota.group,jqfmt=vfsv0,acl  0 2

root@omv:~#uname -a
Linux omv.ad.domain 5.4.0-0.bpo.4-amd64 #1 SMP Debian 5.4.19-1~bpo10+1 (2020-03-09) x86_64 GNU/Linux

root@omv:~# df -h
Filesystem                                                 Size  Used Avail Use% Mounted on
udev                                                       471M     0  471M   0% /dev
tmpfs                                                       99M   15M   85M  15% /run
/dev/sda1                                                  8.8G  2.7G  5.7G  33% /
tmpfs                                                      493M   60K  493M   1% /dev/shm
tmpfs                                                      5.0M     0  5.0M   0% /run/lock
tmpfs                                                      493M     0  493M   0% /sys/fs/cgroup
tmpfs                                                      493M     0  493M   0% /tmp
/dev/sdb1                                                  3.6T  3.6T  3.7G 100% /srv/dev-disk-by-label-data1
/dev/sdd1                                                  3.6T  3.6T     0 100% /srv/dev-disk-by-label-parity1
/dev/sdc1                                                  3.6T  2.1T  1.4T  60% /srv/dev-disk-by-label-data2
overlay                                                    8.8G  2.7G  5.7G  33% /var/lib/docker/overlay2/1ce06260d81726919a513a11cebd62ef96222544f52c91d592f2b0a1f2c26f15/merged
overlay                                                    8.8G  2.7G  5.7G  33% /var/lib/docker/overlay2/5024258ba20d3f223b6933e6b7c294eb47d06b7602b07b056846108edb35667d/merged
shm                                                         64M     0   64M   0% /var/lib/docker/containers/84b230c27c7a878e45f60e1b8b6428878c80b847461690e95c7fb2e3081771a1/mounts/shm
/srv/dev-disk-by-label-data1;/srv/dev-disk-by-label-data2  7.2T  5.7T  1.4T  81% /mnt/data

root@omv:~# cat /etc/samba/smb.conf
[global]
workgroup = SAMDOM
security = ADS
realm = AD.SAMDOM

winbind refresh tickets = Yes
vfs objects = acl_xattr
map acl inherit = Yes
store dos attributes = Yes
hide unreadable = yes
hide unwriteable files = yes

dedicated keytab file = /etc/krb5.keytab
kerberos method = secrets and keytab
winbind enum users = yes
winbind enum groups = yes

log file = /var/log/samba/%m.log
log level = 1 auth:5 winbind:5

idmap config * : backend = tdb
idmap config * : range = 3000-7999
idmap config SAMDOM : backend = rid
idmap config SAMDOM : range = 10000-999999

template shell = /bin/bash
template homedir = /home/%U

username map = /etc/samba/user.map

[nas]
path = /mnt/data/nas
readonly = no

root@omv:~# cat /etc/samba/user.map
!root = SAMDOM\Administrator

root@omv:~# getfacl /mnt/data/nas
getfacl: Removing leading '/' from absolute path names
'# file: mnt/data/nas
'# owner: root
'# group: 1930000512
user::rwx
user:root:rwx
user:10512:rwx
user:10513:rwx
group::---
group:SAMDOM\\domain\040admins:rwx
group:SAMDOM\\domain\040users:rwx
group:1930000512:---
mask::rwx
other::---
default:user::rwx
default:user:root:rwx
default:user:10512:rwx
default:user:10513:rwx
default:group::---
default:group:SAMDOM\\domain\040admins:rwx
default:group:SAMDOM\\domain\040users:rwx
default:group:1930000512:---
default:mask::rwx
default:other::---

**Comment by trapexit** [comment]
Samba leverages xattr and POSIX ACLs for that. There is no "passing through" of Windows ACLs. There's no such thing in POSIX.

xattrs are enabled by default. POSIX ACLs are not. Have you tried with `posix_acl=true`?

You can't compare mhddfs and mergerfs in terms of permissions. It doesn't properly manage permissions in the first place and I'm almost positive that it doesn't support POSIX ACLs given I don't think it was even supported on FUSE at the time it was created.

**Comment by puyolsmash** [comment]
I know I've tried the posix_acl flag before but I guess the permission issue was related to something else... I've set the flag and now it works:
/srv/dev-disk-by-label-data* /mnt/data/ fuse.mergerfs defaults,allow_other,use_ino,posix_acl=true,category.create=mfs 0 0

I think ultimately my issue was that there were still some linux permissions set on the shared folder. Removing those permissions allowed me access to the share. Now it works exactly like I intended, file and folder permissions on the share are completely manageable in Windows. Thnx!
